### PR TITLE
Enable add Claims and AdditionalRequestParameters to request

### PR DESCRIPTION
**Pull request by imcarolwang**
For #4983.

### CODE REVIEW

File: src/System.ServiceModel.Federation/src/System/ServiceModel/Federation/WsTrustTokenParameters.cs [hunk 1]
```diff
@@ -52,8 +52,11 @@ protected WSTrustTokenParameters(WSTrustTokenParameters other)
                 AdditionalRequestParameters.Add(parameter);
 
             CacheIssuedTokens = other.CacheIssuedTokens;
-            foreach (var claimType in ClaimTypes)
-                ClaimTypes.Add(claimType);
+            Claims = other.Claims;
+            foreach(var element in other.AdditionalRequestParameters)
```
Comment (mconnew) [inline src/System.ServiceModel.Federation/src/System/ServiceModel/Federation/WsTrustTokenParameters.cs:56]: You can't just copy the collection over. You have to do a deep copy and copy each item. The reason is that the you don't want changes to the original instance to appear in the copied instance.

File: src/System.ServiceModel.Federation/src/System/ServiceModel/Federation/WSTrustChannelSecurityTokenProvider.cs [hunk 2]
```diff
@@ -167,6 +167,16 @@ protected virtual WsTrustRequest CreateWsTrustRequest()
                 trustRequest.ComputedKeyAlgorithm = _requestSerializationContext.TrustKeyTypes.PSHA1;
             }
 
+            if (WSTrustTokenParameters.Claims != null)
+            {
+                trustRequest.Claims = WSTrustTokenParameters.Claims;
+            }
```
Comment (mconnew) [inline src/System.ServiceModel.Federation/src/System/ServiceModel/Federation/WSTrustChannelSecurityTokenProvider.cs:173]: You need to copy the values over. If someone override this method and modifies the Claims collection, they will end up modifying WSTrustTokenParameters.Claims too.

File: src/System.ServiceModel.Federation/tests/WSTrustChannelSecurityTokenProviderTest.cs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+                TokenType = "urn:oasis:names:tc:SAML:1.0:assertion"
+            };
+
+            tokenRequirement.Properties["http://schemas.microsoft.com/ws/2006/05/servicemodel/securitytokenrequirement/IssuedSecurityTokenParameters"] = tokenParams;
+            tokenRequirement.Properties["http://schemas.microsoft.com/ws/2006/05/servicemodel/securitytokenrequirement/TargetAddress"] = targetAddress;
+            tokenRequirement.Properties["http://schemas.microsoft.com/ws/2006/05/servicemodel/securitytokenrequirement/SecurityAlgorithmSuite"] = System.ServiceModel.Security.SecurityAlgorithmSuite.Default;
+
+            var derivedTokenProvider = new WSTrustChannelSecurityTokenProviderDerived(tokenRequirement);
+            
+            typeof(WSTrustChannelSecurityTokenProvider).GetField("_requestSerializationContext", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance).SetValue(derivedTokenProvider, new WsSerializationContext(WsTrustVersion.TrustFeb2005));
```
Comment (mconnew) [inline src/System.ServiceModel.Federation/tests/WSTrustChannelSecurityTokenProviderTest.cs:62]: >            typeof(WSTrustChannelSecurityTokenProvider).GetField("_requestSerializationContext", System.Reflection.BindingFlags.NonPublic |  [](http://example.com/codeflow?start=0&length=140)

You don't need to do this reflection to set these fields. You can cast the provider to ICommunicationObject and call Open.

File: src/System.ServiceModel.Federation/src/System/ServiceModel/Federation/WsTrustTokenParameters.cs (hunk 1 above)
Comment (imcarolwang) [reply src/System.ServiceModel.Federation/src/System/ServiceModel/Federation/WsTrustTokenParameters.cs]: I see, updated both Claims and AdditionalRequestParameters, could you please help double confirm?

File: src/System.ServiceModel.Federation/src/System/ServiceModel/Federation/WSTrustChannelSecurityTokenProvider.cs (hunk 2 above)
Comment (imcarolwang) [reply src/System.ServiceModel.Federation/src/System/ServiceModel/Federation/WSTrustChannelSecurityTokenProvider.cs:185]: Updated for Claims and AdditionalRequestParameters, could you please help double confirm?

Comment (imcarolwang) [reply src/System.ServiceModel.Federation/src/System/ServiceModel/Federation/WSTrustChannelSecurityTokenProvider.cs:185]: Hi @mconnew, could you please help double confirm if the updated code looks good to you? Thanks!

**Review by mconnew — APPROVED** [review]

**Pull request merged**